### DISCUSSION TITLE
How to get max depth and pollutants and put in a df

### DISCUSSION THREAD

**Discussion by MohammadTahirGhori** [pyswmm/pyswmm#327] [Q&A]
Hello all,
I am working on an optimization process with swmm model. 
I need to get max_report_depth for each specified nodes and put it in a datafram that I will use to calculate flood damage. And also the same way I need pollutants load in links (TSS, TN, TP). 
I used another method of running swmm for my NSGA II algorithm and read the node depths from swmm report file. But it faced an error when I wanted to run for 10000 times. Therefore, I think pyswmm is the best way to get max_report_depth and pollutants load and put it in a datafram. I need to have max depth and pollutants load for every run (10000 run by NSGAII algorithm). 

When I used the introduced code by Pyswmm documentation it gives me all the report for one day and I am using step in advance (900 seconds) then it gives me a list of 95 reads of depths for each nodes.
I need only max depth for each node. 

Can you help me find a way to directly get max_report_depth which is in statistics of nodes and how can I get all max depth for a list of nodes and finally in a df. 

here is what I wrote up to now.

results = []
for node in node_id:
    with Simulation("D:/Thesis/MSc_Yang/strategy1/Basefile.inp") as sim:
        sim.step_advance(900)
        j = Nodes(sim)['ns16']
        for step in sim:
            j.statistics
            results_1 = (j.statistics['max_report_depth'])
            results.append(results_1)
            results.apply(max)
        print (j.statistics['max_report_depth'])

**Comment by MohammadTahirGhori** [comment]
I found this way working but unfortunately it takes a lot of time to compute even one single run. I need to run run swmm about 10000 times. Any way to reduce time or any other method? 
Thanks in advance.


 def node_max_depth(file_path):
     node_depth_dataFrame = pd.DataFrame(columns = ['node_id', 'max_depth'])
     for node in node_id:
         results = []
         with Simulation(file_path) as sim:
             sim.step_advance(900)
             j1 = Nodes(sim)[node]
             # j2 = Nodes(sim)['ns7']
             for step in sim:
                 j1.statistics
                 r = (j1.statistics['max_report_depth'])
                 results.append(r)
             new_row = {'node_id':node, 'max_depth':round(max(results), 2)}
             node_depth_dataFrame = node_depth_dataFrame.append(new_row, ignore_index=True)
     return (node_depth_dataFrame)

**Reply by jennwuu** [reply]
You can try to increase step advance to a larger value if you're using statistics to get your maximum report depth. By increasing the step advance value, pyswmm will pause the swmm engine less frequently.

**Comment by MohammadTahirGhori** [comment]
@jennwuu 
Thanks for your reply>

I increased step advance to read only once the max depth but still it takes long time compared to running 10000 times. 

Can I use Output module to get max depth at nodes and pollutants load at links. 
As I can see in the Output module all codes are with time series. But I need only to know the max depth at nodes and pollutants load at links not at a specific time series.

**Comment by bemcdonnell** [comment]
@MohammadTahirGhori just so you are aware, the max depth in the RPT file is computed on the "routing" time step rather than the "reporting" time step.  Therefore, if you choose to take the max depth from the OUT file you may miss the actual max depth.  Unless you use the `step_advance` feature of PySWMM, PySWMM steps the simulation on the routing step.

What you can do is call that `statistics` function after your simulation rather than during.

**Reply by MohammadTahirGhori** [reply]
Hi @bemcdonnell 
Thanks for your reply. 
Now I am using the simulation and step advance option and I get max depths at nodes but it takes long time while I need to run for 10000. 
You mean I do this 

def nodemax():
    with Simulation ("D:\Thesis\MSc_Yang\strategy1\BaseFile.inp") as sim:
        sim.step_advance(900)
        sim.execute()
        J1 = Nodes(sim)['ns1']
        J1.statistics['max_report_depth']
        print(J1.statistics['max_report_depth'])

I wrote this but it finishes everything and says (restarting the kernel)

About output file 
Is it possible to get max depth and pollutants load. As I read the Output module it is all with timeseries that I do not need.

**Reply by bemcdonnell** [reply]
@MohammadTahirGhori you have two methods in there to run a simulation that are not compatible:

```python
with Simulation(...) as sim:
    for step in sim:
        pass
    J1 = Nodes(sim)['ns1']
    J1.statistics['max_report_depth']
    print(J1.statistics['max_report_depth'])  
```

`sim.execute()`  runs the entire simulation then closes it. Therefore, you would see the restarting kernel.

**Reply by MohammadTahirGhori** [reply]
@bemcdonnell 
Thanks a lot. 

This way it worked for me. 
Now I am using this method to run swmm and directly get max_report_depth as you mentioned. Now when I run for thousands of times it is fast and without any error. 
To confirm,
So I need to run swmm thousands of times and get report of max depth at every run. This method is doing it exactly for me I think. Then I dont need to use sim.execute().
